### ISSUE TITLE
Github Team Roles - Admin needed for 3rd party services

### ISSUE THREAD

**Issue by LeTanque** [Lambda-School-Labs/labs-engineering-standards#53]
Should GH202 specify admin for SL and TL?

> (GH-202) GitHub Team Roles
> GitHub project team membership and team roles will be as follows:
> Section Lead ⇒ Maintainer
> Team Lead ⇒ Maintainer
> Student ⇒ Member
> Rationale:
> Least privilege access
> Exceptions:
> None

Unless teams have admin level access to their repos, they cannot connect to 3rd party services like Heroku.

I think we should clarify that GH202 is referring to Team Roles only and add a section that refers to repository settings. Repository settings use the same taxonomy but with different implications.

**Comment by berniedurfee** [comment]
@LeTanque thanks for the input! You are absolutely correct. We defined org and team roles, but not repo roles... or more precisely the role that a team has on the associated repository.

I've been rethinking this lately, as the issue of requiring admin on a repo to allow 3rd party access comes up often.

For the sake of least privilege, it would be nice to maybe just make the TL an admin, with the rest of the team having push permissions. My main concern being that there are specific configurations that need to remain in place and the chances of those changing increase with more admins on the repo.

With the TL being the only admin, this would require the TL to grant the 3rd party access to the repo. Is that viable?

**Comment by LeTanque** [comment]
@berniedurfee, thanks for the quick response! 

> With the TL being the only admin, this would require the TL to grant the 3rd party access to the repo. Is that viable?

Yes, with some caveats. 

I'll focus on Heroku since it's the 3rd party service that we use the most.

- The TL can connect to Heroku but team members will need access to that heroku instance. We could have TLs add team members as collaborators to the connected Heroku, giving team members the ability to deploy feature branches during development. [Heroku docs](https://devcenter.heroku.com/articles/collaborating)

- In order to make only the TL admin of the team's repo, the TL will need to be in a different 'team'. We could, perhaps, have multiple teams per project (admin team of TL,SL,PL,PM and student team(s)), with each of those teams having their own permissions.


This is assuming, as the documentation suggests, that we are adding students to the github team within the Lambda School Labs organization, and not just as collaborators on the repo. 

In summary: if we give admin access to only the TL, our current Team strategy will have to adjust and we should document the process of adding collaborators to a Heroku project.

**Comment by LeTanque** [comment]
Unless TLs are supposed to only add students as collaborators to projects, and not members of the project team. 

If this is the case, maybe we can describe that in the Standards.

**Comment by tinomen** [comment]
> In order to make only the TL admin of the team's repo, the TL will need to be in a different 'team'.

Can the TL be added to both the team and the repo as an Admin?

**Comment by LeTanque** [comment]
Let preface this with: I am no github expert.  All of my research, thoughts, and conclusions should be mentioned along with "as far as I know". 

The relationship between team and repo _**currently**_ looks something like this:

Lambda Labs Organization
|
|__ Labs PT7 TeamName (each team is a collection of students, TLs, SLs, and PM)
      |
      |___ team-name-fe-pt7
      |___ team-name-be-pt7

Repos can belong to many teams.

Teams can belong to many repos.

Teams are collections of github users.

Users on teams can have different _Lambda Labs Org_ permissions.

Users on teams have the same _repository_ permission. 

> Can the TL be added to both the team and the repo as an Admin?

I believe so.

The TL can be added to both the team as Maintain and repo as an Admin. The TL would be a _collaborator_ and I _think_ that means they can do everything they need to do.

**Comment by berniedurfee** [comment]
The main issue here is that we need to reliably differentiate between roles on a team so we can apply the appropriate permissions. Mission Control will give us a clear picture of those roles, which we can then use to better automate provisioning of GitHub roles. Until this, this standard will be loosely enforced so as not to create more manual tasks.

**Comment by berniedurfee** [comment]
Putting this one on hold until we can automate better. Then we can come up with a structure that make sense as it'll be easy to implement.
